includes/panopto_derivatives.inc

```diff
@@ -5,37 +5,384 @@
+  }
+  if ($response) {
+    $token = $response['access_token'];
+  }
+  return $token;
+}
+
+/**
+ * Retrieves session needed for upload
+ * @param string $token The token retrieved from server
+ * @param string $uploadendpoint The URL which is the target of the request
+ * @return array|null A session retrieved from server
+ */
+function islandora_video_streaming_panopto_get_session($token, $uploadendpoint) {
+  $session = null;
+  $headers = array("Authorization: Bearer {$token}", "Content-Type: application/json");
+  $folder_id = variable_get('islandora_video_streaming_panopto_folder_id');
+  
+  if (isset($folder_id)) {
+    $content = json_encode(array(
+    'FolderId' => $folder_id,
+    ));
+    $response = islandora_video_streaming_panopto_request_curl($uploadendpoint, $headers, $content, 'POST', array(201 => 'ID'));
+    if ($response) {
+      $session = $response;
+    }
+  }
+  return $session;
+}
+
+/**
+ * Function to send a CURL request
+ * @param string $endpoint The URL which is the target of the request
+ * @param array $headers A list of HTTP headers to send
+ * @param string $content The optional body of the request, if applicable
+ * @param string $type The HTTP method to use (defaults to GET)
+ * @param array $expect An optional keyed array of HTTP response codes mapped to an (optional) expected JSON identifier
+ *                      If provided, this is required for the function to return data
+ * @return mixed A string or array.
+ *               If no $expect parameter is given, returns the response body.
+ *               If an $expect key matches with no array value, returns the response body only on a match
+ *               If an $expect key matches and has a value, returns the decoded JSON as an array if that matched value is present as a key
+ */
+function islandora_video_streaming_panopto_request_curl($endpoint, $headers, $content = '', $type = 'GET', $expect = array()) {
+  $cURL = curl_init();
+  curl_setopt($cURL, CURLOPT_URL, $endpoint);
+  curl_setopt($cURL, CURLOPT_RETURNTRANSFER, true);
+  curl_setopt($cURL, CURLOPT_HEADER, 1);
+  curl_setopt($cURL, CURLINFO_HEADER_OUT, true);
+  switch (strtoupper($type)) {
+    case 'POST':
+    curl_setopt($cURL, CURLOPT_POST, true);
+    break;
+    case 'PUT':
+    curl_setopt($cURL, CURLOPT_CUSTOMREQUEST, 'PUT');
+    break;
+  }
+  if ($content) {
+    curl_setopt($cURL, CURLOPT_POSTFIELDS, $content);
+  }
+  curl_setopt($cURL, CURLOPT_HTTPHEADER, $headers);
+  $response = curl_exec($cURL);
+  $header_size = curl_getinfo($cURL, CURLINFO_HEADER_SIZE);
+  $response_code = curl_getinfo($cURL, CURLINFO_RESPONSE_CODE);
+  $headerSent = curl_getinfo($cURL, CURLINFO_HEADER_OUT );
+  curl_close($cURL);
+  if ($response === false) {
+    $messages = array(
+      'message' => t("Unable to receive response from curl_exec()"),
+      'message_sub' => array(),
```

Comment:
Use `curl_error($cURL)` to fetch the last error(s) recorded by the cURL library for this `$cURL` object, to make the error message here more meaningful.  Pass the curl errors to "message" via a "message_sub".